controllers/upgrade_controller.go

```diff
@@ -0,0 +1,270 @@
+			newObj, _ := e.ObjectNew.(*v1alpha1.StorageClient)
+			if oldObj == nil || newObj == nil {
+				return false
+			}
+
+			if newObj.Status.Operator.CurrentVersion != newObj.Status.Operator.DesiredVersion {
+				return true
+			}
+
+			return false
+		},
+	}
+	ocsClientOperatorSubscriptionPredicate := predicate.Funcs{
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			if e.ObjectOld == nil || e.ObjectNew == nil {
+				return false
+			}
+
+			oldObj, _ := e.ObjectOld.(*opv1a1.Subscription)
+			newObj, _ := e.ObjectNew.(*opv1a1.Subscription)
+			if oldObj == nil || newObj == nil {
+				return false
+			}
+
+			if oldObj.Spec.Package != ocsClientOperatorSubscriptionPackageName {
+				return false
+			}
+
+			return oldObj.Spec.Channel != newObj.Spec.Channel
+		},
+	}
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(controllerName).
+		Watches(&v1alpha1.StorageClient{}, &handler.EnqueueRequestForObject{}, builder.WithPredicates(storageClientOperatorVersionPredicate)).
+		Watches(&opv1a1.Subscription{}, &handler.EnqueueRequestForObject{}, builder.WithPredicates(ocsClientOperatorSubscriptionPredicate)).
+		Complete(r)
+}
+
+//+kubebuilder:rbac:groups=operators.coreos.com,resources=subscriptions,verbs=get;list;patch
+
+func (r *UpgradeReconciler) Reconcile(ctx context.Context, _ ctrl.Request) (ctrl.Result, error) {
+	r.ctx = ctx
+	r.log = log.FromContext(ctx)
+	r.log.Info("starting reconcile")
+
+	conditionStatus := metav1.ConditionTrue
+	conditionMessage := "No component reported errors."
+	result, err := r.reconcilePhases()
+	if err != nil {
+		// TODO (lgangava): we are being very restrictive as of now, not upgrading if we hit any error
+		// differentiate between reconciler errors and not upgradeable errors
+		conditionStatus = metav1.ConditionFalse
+		conditionMessage = fmt.Sprintf("Operator could not be upgraded due to: %v", err)
+		r.log.Error(err, "an error occurred during reconcilePhases")
+	}
+	err = errors.Join(err, r.setOperatorCondition(conditionStatus, conditionMessage))
+
+	r.log.Info("reconciling completed")
+	return result, err
+}
+
+func (r *UpgradeReconciler) reconcilePhases() (ctrl.Result, error) {
+	if err := r.reconcileSubscriptionChannel(); err != nil {
+		return ctrl.Result{}, err
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *UpgradeReconciler) reconcileSubscriptionChannel() error {
+	storageClientList := &v1alpha1.StorageClientList{}
+	if err := r.list(storageClientList); err != nil {
+		return fmt.Errorf("failed to list storageclients %v", err)
+	}
+
+	if len(storageClientList.Items) == 0 {
+		r.log.Info("no storageclients exist")
+		return nil
+	}
+
+	desiredVersions := make([]string, 0, len(storageClientList.Items))
+	for idx := range storageClientList.Items {
+		desiredVersion := storageClientList.Items[idx].Status.Operator.DesiredVersion
+		// usually these versions will not have patch version, so we are adding an arbitrary num for semver comparison
+		desiredVersion += dummyPatchVersion
```

Comment:
Doesn't not having a patch version translate to `_._.0` and not some arbitrary number?

Reply:
nope, `0` is as good as any other random number and we just need a random number.
the background is:
- provider sends desiredversion and it'll have only Major.minor
- client queries it's version and sets currentVersion, also Major.minor
- but semver is strict about version be Major.minor.patch and for comparison we just need a constant random number added to it.

why not `0` then?
- if this code changes and it's logged, M.m.0 will actually be a release version but the release to be M.m.99 is slim to none so I chose 99